numpyro/distributions/util.py

```diff
@@ -44,27 +46,74 @@ def categorical(key, p, shape=()):
     return _categorical(key, p, shape)
 
 
+# Ref https://github.com/numpy/numpy/blob/8a0858f3903e488495a56b4a6d19bbefabc97dca/
+# numpy/random/src/distributions/distributions.c#L574
+def _poisson_large(val):
+    rng_key, lam = val
+    slam = np.sqrt(lam)
+    loglam = np.log(lam)
+    b = 0.931 + 2.53 * slam
+    a = -0.059 + 0.02483 * b
+    invalpha = 1.1239 + 1.1328 / (b - 3.4)
+    vr = 0.9277 - 3.6224 / (b - 2)
+
+    def cond_fn(val):
+        _, V, us, k = val
+        cond1 = (us >= 0.07) & (V <= vr)
+        cond2 = (k < 0) | ((us < 0.013) & (V > us))
+        cond3 = ((np.log(V) + np.log(invalpha) - np.log(a / (us * us) + b))
+                 <= (-lam + k * loglam - gammaln(k + 1)))
+        return (~cond1) & (cond2 | (~cond3))
+
+    def body_fn(val):
+        rng_key, *_ = val
+        rng_key, key_U, key_V = random.split(rng_key, 3)
+        U = random.uniform(key_U) - 0.5
+        V = random.uniform(key_V)
+        us = 0.5 - np.abs(U)
+        k = np.floor((2 * a / us + b) * U + lam + 0.43)
+        return rng_key, V, us, k
+
+    *_, k = lax.while_loop(cond_fn, body_fn, (rng_key, 0., 0., -1.))
+    return k
+
+
+def _poisson_small(val):
+    rng_key, lam = val
+    enlam = np.exp(-lam)
+
+    def body_fn(val):
+        rng_key, prod, k = val
+        rng_key, key_U = random.split(rng_key)
+        U = random.uniform(key_U)
+        prod = prod * U
+        return rng_key, prod, k + 1
+
+    init = np.where(lam == 0., 0., -1.)
+    *_, k = lax.while_loop(lambda val: val[1] > enlam, body_fn, (rng_key, 1., init))
+    return k
+
+
+def _poisson_one(val):
+    return lax.cond(val[1] >= 10, val, _poisson_large, val, _poisson_small)
+
+
 @partial(jit, static_argnums=(2, 3))
 def _poisson(key, rate, shape, dtype):
     # Ref: https://en.wikipedia.org/wiki/Poisson_distribution#Generating_Poisson-distributed_random_variables
     shape = shape or np.shape(rate)
-    L = np.exp(-rate)
-    k = np.zeros(shape, dtype=dtype)
-    p = np.ones(shape)
-
-    def body_fn(val):
-        k, p, rng_key = val
-        k = np.where(p > L, k + 1, k)
-        rng_key, rng_key_u = random.split(rng_key)
-        u = random.uniform(rng_key_u, shape)
-        p = p * u
-        return k, p, rng_key
-
-    k, _, _ = lax.while_loop(lambda val: np.any(val[1] > L), body_fn, (k, p, key))
-    return k - 1
+    rate = lax.convert_element_type(rate, canonicalize_dtype(np.float64))
+    rate = np.broadcast_to(rate, shape)
+    rng_keys = random.split(key, np.size(rate))
+    if xla_bridge.get_backend().platform == 'cpu':
```

Comment:
I am curious as to why this should differ with the backend.

Reply:
For rejective samplers, it is better to use `lax.map` because in the vectorized map, early-finished samples have to wait for other proposals accepted. However, in GPU vectorized is preferred (i.e. faster).